File: multi-lang/python/bench_contract.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
QZDB reference-compliant benchmark (Python) — implements docs/BENCH_CONTRACT.md.

This is the TEMPLATE the other 5 non-compliant languages (C/Go/Rust/Node/PHP)
replicate. It:
  1. imports gen_stream()/enc_query() from tools/bench_gen.py (single source
     of truth) and SELF-CHECKS parity: the first-1024 SHA of every stream it
     generates must equal bench_vectors.json's fingerprint, else it aborts.
  2. runs 4 distributions x 3 dual-stack modes with QPS / avg / P50 / P95 / P99,
     cold vs hot, and 1/8/16-thread scaling (Python is GIL-bound, so thread
     QPS is reported best-effort + a concurrency-safety assertion).
  3. adds a `hot.mixed` string round-trip to isolate parse vs decode cost.
  4. writes the canonical JSON to multi-lang/bench_reports/python_<edition>.json.

Env overrides (for quick local runs):
    BENCH_OPS=200000          scale OPS down
    BENCH_EDITIONS=std_china  comma list of editions to run
"""

import hashlib
import json
import os
import platform
import statistics
import sys
import threading
import time

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)
TOOLS = os.path.join(HERE, "..", "tools")
sys.path.insert(0, TOOLS)
REPORTS = os.path.join(HERE, "..", "bench_reports")

import bench_gen as bg
from qzdb import QzdbReader  # reference SDK under test

LAT_SAMPLE_EVERY = 20
THREAD_CONFIGS = [1, 8, 16]
CONCURRENCY_THREADS = 16
CONCURRENCY_OPS = 100_000

EDITIONS = {
    "std_china": ("std", "china", "qqzeng_ip_std_china.qzdb"),
    "max_global": ("max", "global", "qqzeng_ip_max_global.qzdb"),
}

DB_ROOT_CANDIDATES = [
    os.path.join(HERE, "..", "test_data_202608"),
    os.path.join(HERE, "test_data_202608"),
    os.path.join(HERE, "..", "..", "test_data_202608"),
    os.path.join(HERE, "data"),
]


def find_db(edition: str) -> str:
    reg, region, fn = EDITIONS[edition]
    for root in DB_ROOT_CANDIDATES:
        p = os.path.join(root, reg, region, fn)
        if os.path.exists(p):
            return p
    return None


def fmt_v4(ip: int) -> str:
    return f"{(ip >> 24) & 255}.{(ip >> 16) & 255}.{(ip >> 8) & 255}.{ip & 255}"


def fmt_v6(h: int, l: int) -> str:
    v = (h << 64) | l
    return ":".join(f"{(v >> (112 - 16 * k)) & 0xFFFF:x}" for k in range(8))


def dispatch(reader, q):
    """Route each query kind to the entry point a real caller would use.

    'm' (IPv4-mapped, ::ffff:w.x.y.z) MUST go through the mapped-aware bytes
    API: find_v6_uint() is a pure v6 trie walk that never downgrades, so it
    reports a miss for every mapped query and the bench ends up timing the
    early-exit miss path instead of a real lookup. Keeping the three kinds on
    three distinct entry points is what makes the numbers comparable across
    languages -- see docs/BENCH_CONTRACT.md.
    """
    if q[0] == "v4":
        return reader.find_uint(q[1])
    if q[0] == "m":
        return reader.find_bytes(((q[1] << 64) | q[2]).to_bytes(16, "big"))
    return reader.find_v6_uint((q[1] << 64) | q[2])


def percentile(ns_list, p):
    if not ns_list:
        return 0
    ns_list.sort()
    idx = min(len(ns_list) - 1, max(0, int(__import__("math").ceil(p * len(ns_list))) - 1))
    return ns_list[idx]


def run_single(reader, dist, mode, seed, pool_v4, pool_v6, ops, sample=True):
    lat = []
    hits = 0
    start = time.perf_counter()
    for i, q in enumerate(bg.gen_stream(dist, mode, seed, pool_v4, pool_v6)):
        if sample and i % LAT_SAMPLE_EVERY == 0:
            t0 = time.perf_counter()
            found = dispatch(reader, q)
            lat.append(int((time.perf_counter() - t0) * 1e9))
        else:
            found = dispatch(reader, q)
        if found is not None:
            hits += 1
        if i + 1 >= ops:
            break
    elapsed = time.perf_counter() - start
    qps = ops / elapsed
    return {
        "ops": ops,
        "qps": round(qps, 1),
        "avg_ns": round(elapsed * 1e9 / ops, 1),
        "p50_ns": int(percentile(lat, 0.50)) if lat else 0,
        "p95_ns": int(percentile(lat, 0.95)) if lat else 0,
        "p99_ns": int(percentile(lat, 0.99)) if lat else 0,
        "errors": 0,
        # hit_rate is mandatory: a QPS number without it is uninterpretable,
        # because a fast run may simply be one where every query missed and
        # took the early-exit path.
        "hits": hits,
        "hit_rate": hits / ops,
    }


def run_multi(reader, dist, mode, seed, pool_v4, pool_v6, threads, ops):
    per = ops // threads
    errs = [0]
    done = [0]
    latch = threading.Barrier(threads)
    start = [0.0]

    def worker(base):
        try:
            latch.wait()
            if base == 0:
                start[0] = time.perf_counter()
            for i, q in enumerate(bg.gen_stream(dist, mode, seed, pool_v4, pool_v6)):
                dispatch(reader, q)
                done[0] += 1
                if i + 1 >= per:
                    break
        except Exception:
            errs[0] += 1

    ts = [threading.Thread(target=worker, args=(t * per,)) for t in range(threads)]
    for t in ts:
        t.start()
    for t in ts:
        t.join()
    elapsed = time.perf_counter() - start[0]
    return {
        "ops": done[0],
        "qps": round(done[0] / elapsed, 1),
        "avg_ns": round(elapsed * 1e9 / max(1, done[0]), 1),
        "p50_ns": 0, "p95_ns": 0, "p99_ns": 0,
        "errors": errs[0],
    }


def concurrency_safe(reader, dist, mode, seed, pool_v4, pool_v6):
    per = CONCURRENCY_OPS // CONCURRENCY_THREADS
    errs = [0]
    done = [0]

    def worker(base):
        try:
            for i, q in enumerate(bg.gen_stream(dist, mode, seed, pool_v4, pool_v6)):
                dispatch(reader, q)
                done[0] += 1
                if i + 1 >= per:
                    break
        except Exception:
            errs[0] += 1

    ts = [threading.Thread(target=worker, args=(t * per,)) for t in range(CONCURRENCY_THREADS)]
    for t in ts:
        t.start()
    for t in ts:
        t.join()
    return errs[0] == 0 and done[0] == CONCURRENCY_THREADS * per


def parity_selfcheck(manifest):
    print("parity self-check ...", end=" ", flush=True)
    pool_v4, pool_v6 = bg.build_pools()
    bad = 0
    for dist, modes in manifest["streams"].items():
        for mode, info in modes.items():
            h = hashlib.sha256()
            for i, q in enumerate(bg.gen_stream(dist, mode, info["seed"], pool_v4, pool_v6)):
                if i >= bg.FINGERPRINT_N:
                    break
                h.update(bg.enc_query(q))
            if h.hexdigest() != info["first1024_sha256"]:
                bad += 1
                print(f"\n  MISMATCH {dist}.{mode}!")
    if bad:
        print("FAILED")
        sys.exit(1)
    print("OK (12/12 streams match bench_vectors.json)")


def main():
    ops = int(os.environ.get("BENCH_OPS", bg.OPS))
    editions = os.environ.get("BENCH_EDITIONS", "std_china,max_global").split(",")
    manifest = json.load(open(os.path.join(TOOLS, "bench_vectors.json"), encoding="utf-8"))
    parity_selfcheck(manifest)
    pool_v4, pool_v6 = bg.build_pools()

    os.makedirs(REPORTS, exist_ok=True)

    for edition in editions:
        db = find_db(edition)
        if not db:
            print(f"[SKIP] {edition}: db not found")
            continue
        reader = QzdbReader(db)
        print(f"\nedition {edition}: {db} ({os.path.getsize(db):,} bytes)")

        # concurrency safety on hot.mixed (most representative)
        safe = concurrency_safe(reader, "hot", "mixed",
                                manifest["streams"]["hot"]["mixed"]["seed"], pool_v4, pool_v6)
        print(f"  concurrency_safe(16x{CONCURRENCY_OPS//1000}k): {safe}")

        dist_out = {}
        for dist, modes in manifest["streams"].items():
            dist_out[dist] = {}
            for mode, info in modes.items():
                seed = info["seed"]

                # cold: first COLD_OPS right after load (cache cold)
                cold = run_single(reader, dist, mode, seed, pool_v4, pool_v6,
                                  min(ops, bg.COLD_OPS), sample=True)
                cold["warm"] = "cold"

                # warmup
                run_single(reader, dist, mode, seed, pool_v4, pool_v6,
                           min(ops, bg.WARMUP_OPS), sample=False)

                # hot: single-thread full ops
                hot = run_single(reader, dist, mode, seed, pool_v4, pool_v6, ops, sample=True)
                hot["warm"] = "hot"

                threads = {}
                for tc in THREAD_CONFIGS:
                    r = run_multi(reader, dist, mode, seed, pool_v4, pool_v6, tc, ops)
                    r["warm"] = "hot"
                    threads[str(tc)] = r

                dist_out[dist][mode] = {"cold": cold, "hot": hot, "threads": threads}

                print(f"  {dist:11s}.{mode:6s}  hot QPS={hot['qps']:>12,.0f}  "
                      f"p50={hot['p50_ns']:>6d}ns  p99={hot['p99_ns']:>7d}ns  "
                      f"8T={threads['8']['qps']:>12,.0f}  err={hot['errors']}  "
                      f"hit={hot['hit_rate'] * 100:.1f}%")

        # string round-trip on hot.mixed
        seed = manifest["streams"]["hot"]["mixed"]["seed"]
        lat = []
        start = time.perf_counter()
        n = 0
        for i, q in enumerate(bg.gen_stream("hot", "mixed", seed, pool_v4, pool_v6)):
            s = fmt_v4(q[1]) if q[0] == "v4" else fmt_v6(q[1], q[2])
            if i % LAT_SAMPLE_EVERY == 0:
                t0 = time.perf_counter()
                reader.find(s)
                lat.append(int((time.perf_counter() - t0) * 1e9))
            else:
                reader.find(s)
            n += 1
            if n >= ops:
                break
        elapsed = time.perf_counter() - start
        string_rt = {
            "api": "string", "ops": n, "qps": round(n / elapsed, 1),
            "avg_ns": round(elapsed * 1e9 / n, 1),
            "p50_ns": int(percentile(lat, 0.50)) if lat else 0,
            "p95_ns": int(percentile(lat, 0.95)) if lat else 0,
            "p99_ns": int(percentile(lat, 0.99)) if lat else 0, "errors": 0, "warm": "hot",
        }
        print(f"  {'hot':11s}.{'mixed':6s}  STRING round-trip QPS={string_rt['qps']:>12,.0f}  "
              f"p99={string_rt['p99_ns']:>7d}ns")

        report = {
            "contract": "QZDB_BENCH_CONTRACT v1.0",
            "language": "python",
            "sdk_version": "multi-lang/python (reference)",
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%S%z"),
            "seed": bg.MASTER_SEED,
            "db": {"path": db, "edition": edition,
                   "bytes": os.path.getsize(db),
                   "hash": f"crc32:n/a"},  # Python SDK exposes getFileHash on reader if available
            "environment": {
                "cpu": platform.processor() or platform.machine(),
                "cores": os.cpu_count(),
                "ram_gb": None,
                "os": f"{platform.system()} {platform.release()} {platform.machine()}",
                "runtime": f"Python {platform.python_version()}",
                "compiler": "CPython",
                "bench_contract": "v1.0",
                "note": "Python is GIL-bound; multi-thread QPS is best-effort, "
                        "concurrency_safety is the meaningful thread assertion.",
            },
            "distributions": dist_out,
            "string_roundtrip": {"hot": {"mixed": string_rt}},
            "concurrency_safe": safe,
        }
        out_path = os.path.join(REPORTS, f"python_{edition}.json")
        with open(out_path, "w", encoding="utf-8") as f:
            json.dump(report, f, indent=2, ensure_ascii=False)
        print(f"  wrote {out_path}")


if __name__ == "__main__":
    main()

File: multi-lang/nodejs/bench_contract.js
#!/usr/bin/env node
'use strict';
/*
 * QZDB reference-compliant benchmark (Node.js) — docs/BENCH_CONTRACT.md
 * Ports gen_stream() from tools/bench_gen.py and SELF-CHECKS FNV-1a parity
 * against bench_vectors.json. Template for the other languages.
 */
const fs = require('fs');
const path = require('path');
const os = require('os');
const { Worker, isMainThread } = require('worker_threads');
const QzdbReader = require('./qzdb.js');

const MASK = (1n << 64n) - 1n;
const MASTER_SEED = 20260807n;
const OPS_FULL = 2000000;
const POOL_V4 = 4096, POOL_V6 = 1024;
const FINGERPRINT_N = 1024;
const LAT_EVERY = 20;

function fnv1a(buf, h = 0n) {
  if (h === 0n) h = 0xCBF29CE484222325n;
  for (let i = 0; i < buf.length; i++) {
    h ^= BigInt(buf[i]);
    h = (h * 0x100000001B3n) & MASK;
  }
  return h;
}

class SM {
  constructor(seed) { this.s = BigInt(seed) & MASK; }
  next() {
    this.s = (this.s + 0x9E3779B97F4A7C15n) & MASK;
    let z = this.s;
    z = ((z ^ (z >> 30n)) * 0xBF58476D1CE4E5B9n) & MASK;
    z = ((z ^ (z >> 27n)) * 0x94D049BB133111EBn) & MASK;
    return z ^ (z >> 31n);
  }
  u32() { return this.next() & 0xFFFFFFFFn; }
  u128() { return ((this.next() & MASK) << 64n) | (this.next() & MASK); }
}

function buildPools() {
  const p4 = new SM(MASTER_SEED + 1n), pool_v4 = [];
  for (let i = 0; i < POOL_V4; i++) pool_v4.push(p4.u32());
  const p6 = new SM(MASTER_SEED + 2n), pool_v6 = [];
  for (let i = 0; i < POOL_V6; i++) pool_v6.push(p6.u128());
  return [pool_v4, pool_v6];
}

function genV4(dist, rng, pool_v4, base4, i) {
  if (dist === 'random') return rng.u32();
  if (dist === 'hot') return pool_v4[Number(rng.u32() % 4096n)];
  if (dist === 'sequential') return (base4 + BigInt(i)) & 0xFFFFFFFFn;
  const r = rng.u32() % 10n;
  if (r < 6n) return pool_v4[Number(rng.u32() % 4096n)];
  if (r < 9n) return rng.u32();
  return (base4 + BigInt(i)) & 0xFFFFFFFFn;
}
function genV6(dist, rng, pool_v6, base6, i) {
  if (dist === 'random') { const v = rng.u128(); return [v >> 64n, v & MASK]; }
  if (dist === 'hot') { const v = pool_v6[Number(rng.u32() % 1024n)]; return [v >> 64n, v & MASK]; }
  if (dist === 'sequential') { const v = (base6 + BigInt(i)) & ((1n << 128n) - 1n); return [v >> 64n, v & MASK]; }
  const r = rng.u32() % 10n;
  if (r < 6n) { const v = pool_v6[Number(rng.u32() % 1024n)]; return [v >> 64n, v & MASK]; }
  if (r < 9n) { const v = rng.u128(); return [v >> 64n, v & MASK]; }
  const v = (base6 + BigInt(i)) & ((1n << 128n) - 1n); return [v >> 64n, v & MASK];
}
function mapped(ip) {
  const v = (0xFFFFn << 32n | (ip & 0xFFFFFFFFn)) & ((1n << 128n) - 1n);
  return [v >> 64n, v & MASK];
}
function encV4(ip) { const b = Buffer.alloc(8); b.writeBigUInt64LE(ip & 0xFFFFFFFFn, 0); return b; }
function encV6(h, l) { const b = Buffer.alloc(16); b.writeBigUInt64LE(h & MASK, 0); b.writeBigUInt64LE(l & MASK, 8); return b; }

function* genStream(dist, mode, seed, pool_v4, pool_v6) {
  const rng = new SM(seed);
  const base4 = rng.u32();
  const base6 = rng.u128();
  for (let i = 0; i < OPS_FULL; i++) {
    if (mode === 'v4') yield [0, genV4(dist, rng, pool_v4, base4, i), 0n];
    else if (mode === 'v6') {
      if (rng.u32() % 5n === 0n) { const [h, l] = mapped(genV4(dist, rng, pool_v4, base4, i)); yield [2, h, l]; }
      else { const [h, l] = genV6(dist, rng, pool_v6, base6, i); yield [1, h, l]; }
    } else {
      const m = i % 10;
      if (m < 5) yield [0, genV4(dist, rng, pool_v4, base4, i), 0n];
      else if (m < 9) { const [h, l] = genV6(dist, rng, pool_v6, base6, i); yield [1, h, l]; }
      else { const [h, l] = mapped(genV4(dist, rng, pool_v4, base4, i)); yield [2, h, l]; }
    }
  }
}
function encQuery(q) { return q[0] === 0 ? encV4(q[1]) : encV6(q[1], q[2]); }

function percentile(arr, p) {
  if (!arr.length) return 0;
  const a = arr.slice().sort((x, y) => x - y);
  const idx = Math.min(a.length - 1, Math.max(0, Math.ceil(p * a.length) - 1));
  return a[idx];
}

function findDb(edition, reg, region, fn) {
  const roots = [
    path.join(__dirname, '..', 'test_data_202608'),
    path.join(__dirname, 'test_data_202608'),
    path.join(__dirname, '..', '..', 'test_data_202608'),
  ];
  for (const r of roots) { const p = path.join(r, reg, region, fn); if (fs.existsSync(p)) return p; }
  return null;
}

// Route each query kind to the entry point a real caller would use, and return
// whether it hit. kind 2 is IPv4-mapped (::ffff:w.x.y.z) and MUST use
// findBytes(), which performs the mapped downgrade; findV6() is a pure v6 trie
// walk that never downgrades, so it would miss on every mapped query and the
// bench would time the early-exit miss path instead of a real lookup.
// `buf` is a caller-owned scratch Buffer — findBytes() does not retain it.
function dispatch(reader, q, buf) {
  if (q[0] === 0) return reader.findUint(Number(q[1])) !== null;
  if (q[0] === 2) {
    buf.writeBigUInt64BE(q[1], 0);
    buf.writeBigUInt64BE(q[2], 8);
    return reader.findBytes(buf) !== null;
  }
  return reader.findV6(q[1], q[2]) !== null;
}

function runSingle(reader, dist, mode, seed, pool_v4, pool_v6, ops, sample) {
  const lat = [];
  const buf = Buffer.allocUnsafe(16);
  let hits = 0;
  const t0 = process.hrtime.bigint();
  let i = 0;
  for (const q of genStream(dist, mode, seed, pool_v4, pool_v6)) {
    let found;
    if (sample && i % LAT_EVERY === 0) {
      const s = process.hrtime.bigint();
      found = dispatch(reader, q, buf);
      lat.push(Number(process.hrtime.bigint() - s));
    } else {
      found = dispatch(reader, q, buf);
    }
    if (found) hits++;
    if (++i >= ops) break;
  }
  const elapsed = Number(process.hrtime.bigint() - t0) / 1e9;
  // hit_rate is reported because a QPS number without it is uninterpretable:
  // a high QPS may just mean every query took the early-exit miss path.
  return { ops, qps: Math.round(ops / elapsed), avg_ns: Math.round(elapsed * 1e9 / ops),
    p50_ns: percentile(lat, 0.5), p95_ns: percentile(lat, 0.95), p99_ns: percentile(lat, 0.99),
    errors: 0, hits, hit_rate: hits / ops };
}

function parityCheck(manifest, pool_v4, pool_v6) {
  let bad = 0;
  for (const dist of Object.keys(manifest.streams)) {
    for (const mode of Object.keys(manifest.streams[dist])) {
      const info = manifest.streams[dist][mode];
      let fnv = 0n;
      let i = 0;
      for (const q of genStream(dist, mode, info.seed, pool_v4, pool_v6)) {
        if (i++ >= FINGERPRINT_N) break;
        fnv = fnv1a(encQuery(q), fnv);
      }
      if (fnv !== BigInt(String(info.first1024_fnv1a))) { bad++; console.log('  MISMATCH', dist, mode); }
    }
  }
  return bad === 0;
}

function runWorkerQueries(db, dist, mode, seed, count) {
  const reader = new QzdbReader(db, 0, false);
  const [pool_v4, pool_v6] = buildPools();
  const buf = Buffer.allocUnsafe(16);
  let done = 0, err = 0;
  for (const q of genStream(dist, mode, seed, pool_v4, pool_v6)) {
    try { dispatch(reader, q, buf); done++; }   // a miss is expected, not a failure
    catch (e) { err++; break; }
    if (done >= count) break;
  }
  return { done, err };
}

const { spawn } = require('child_process');
function concurrencySafe(db, dist, mode, seed) {
  const n = 16, per = 100000 / n;
  return new Promise((resolve) => {
    let pending = n, errors = 0, done = 0;
    const self = path.join(__dirname, 'bench_contract.js');
    for (let t = 0; t < n; t++) {
      const cp = spawn(process.execPath, ['-e',
        `process.env.BENCH_WORKER='1';` +
        `try{const r=require(${JSON.stringify(self)}).runWorkerQueries(${JSON.stringify(db)},${JSON.stringify(dist)},${JSON.stringify(mode)},${seed},${per});console.log(JSON.stringify(r));process.exit(0)}catch(e){console.error(String(e&&e.stack||e));process.exit(1)}`
      ], { stdio: ['ignore', 'pipe', 'inherit'] });
      let buf = '';
      cp.stdout.on('data', (d) => buf += d);
      cp.on('close', (code) => {
        if (code === 0) { const r = JSON.parse(buf); done += r.done; errors += r.err; }
        else errors++;
        if (--pending === 0) resolve(errors === 0 && done === n * per);
      });
    }
  });
}

async function main() {
  const ops = parseInt(process.env.BENCH_OPS || OPS_FULL, 10);
  const editions = (process.env.BENCH_EDITIONS || 'std_china,max_global').split(',');
  const manifest = JSON.parse(fs.readFileSync(path.join(__dirname, '..', 'tools', 'bench_vectors.json'), 'utf8'));
  const [pool_v4, pool_v6] = buildPools();
  process.stdout.write('parity self-check ... ');
  if (!parityCheck(manifest, pool_v4, pool_v6)) { console.log('FAILED'); process.exit(1); }
  console.log('OK (12/12)');

  const reportsDir = path.join(__dirname, '..', 'bench_reports');
  fs.mkdirSync(reportsDir, { recursive: true });
  const editionsMap = { std_china: ['std', 'china', 'qqzeng_ip_std_china.qzdb'], max_global: ['max', 'global', 'qqzeng_ip_max_global.qzdb'] };

  for (const edition of editions) {
    const [reg, region, fn] = editionsMap[edition];
    const db = findDb(edition, reg, region, fn);
    if (!db) { console.log(`[SKIP] ${edition}: not found`); continue; }
    const reader = new QzdbReader(db, 0, false);
    console.log(`\nedition ${edition}: ${db} (${fs.statSync(db).size} bytes)`);

    const safe = await concurrencySafe(db, 'hot', 'mixed', manifest.streams.hot.mixed.seed, pool_v4, pool_v6);
    console.log(`  concurrency_safe(16x100k): ${safe}`);

    const distOut = {};
    for (const dist of Object.keys(manifest.streams)) {
      distOut[dist] = {};
      for (const mode of Object.keys(manifest.streams[dist])) {
        const seed = manifest.streams[dist][mode].seed;
        const cold = runSingle(reader, dist, mode, seed, pool_v4, pool_v6, Math.min(ops, 200000), true); cold.warm = 'cold';
        runSingle(reader, dist, mode, seed, pool_v4, pool_v6, Math.min(ops, 1000000), false);
        const hot = runSingle(reader, dist, mode, seed, pool_v4, pool_v6, ops, true); hot.warm = 'hot';
        distOut[dist][mode] = { cold, hot, threads: { '1': hot } };
        console.log(`  ${dist.padEnd(11)}.${mode.padEnd(6)} hot QPS=${String(hot.qps).padStart(12)} p50=${String(hot.p50_ns).padStart(6)}ns p99=${String(hot.p99_ns).padStart(7)}ns err=${hot.errors} hit=${(hot.hit_rate * 100).toFixed(1)}%`);
      }
    }

    // string round-trip on hot.mixed
    const seed = manifest.streams.hot.mixed.seed;
    const lat = []; let n = 0; const t0 = process.hrtime.bigint();
    for (const q of genStream('hot', 'mixed', seed, pool_v4, pool_v6)) {
      const s = q[0] === 0 ? `${(Number(q[1]) >> 24) & 255}.${(Number(q[1]) >> 16) & 255}.${(Number(q[1]) >> 8) & 255}.${Number(q[1]) & 255}`
        : (() => { const v = (q[1] << 64n) | q[2]; const g = []; for (let k = 0; k < 8; k++) g.push(((v >> (112n - 16n * BigInt(k))) & 0xFFFFn).toString(16)); return g.join(':'); })();
      if (n % LAT_EVERY === 0) { const ss = process.hrtime.bigint(); reader.find(s); lat.push(Number(process.hrtime.bigint() - ss)); }
      else reader.find(s);
      if (++n >= ops) break;
    }
    const el = Number(process.hrtime.bigint() - t0) / 1e9;
    const string_rt = { api: 'string', ops: n, qps: Math.round(n / el), avg_ns: Math.round(el * 1e9 / n), p50_ns: percentile(lat, 0.5), p95_ns: percentile(lat, 0.95), p99_ns: percentile(lat, 0.99), errors: 0, warm: 'hot' };
    console.log(`  ${'hot'.padEnd(11)}.${'mixed'.padEnd(6)} STRING round-trip QPS=${String(string_rt.qps).padStart(12)} p99=${String(string_rt.p99_ns).padStart(7)}ns`);

    const report = {
      contract: 'QZDB_BENCH_CONTRACT v1.0', language: 'nodejs', sdk_version: 'multi-lang/nodejs',
      timestamp: new Date().toISOString(), seed: Number(MASTER_SEED),
      db: { path: db, edition, bytes: fs.statSync(db).size, hash: 'crc32:n/a' },
      environment: { cpu: os.cpus()[0] ? os.cpus()[0].model : '?', cores: os.cpus().length, ram_gb: Math.round(os.totalmem() / 1e9), os: `${os.type()} ${os.release()}`, runtime: `Node ${process.version}`, compiler: 'V8', bench_contract: 'v1.0', note: 'Node is single-threaded; concurrency_safety uses worker_threads (per-worker reader).' },
      distributions: distOut, string_roundtrip: { hot: { mixed: string_rt } }, concurrency_safe: safe,
    };
    const out = path.join(reportsDir, `nodejs_${edition}.json`);
    fs.writeFileSync(out, JSON.stringify(report, null, 2));
    console.log(`  wrote ${out}`);
  }
}
const isEntry = (typeof require !== 'undefined') && require.main === module;
if (isEntry && !process.env.BENCH_WORKER) main();

module.exports = { runWorkerQueries, genStream, buildPools, parityCheck };
